// office-ui-fabric-react/lib/components/Facepile/FacepilePage.visualtest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var index_1 = require("./index");
/* tslint:disable:no-unused-variable */
var React = require("react");
/* tslint:enable:no-unused-variable */
var FacepileVPage = (function (_super) {
    tslib_1.__extends(FacepileVPage, _super);
    function FacepileVPage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FacepileVPage.prototype.render = function () {
        return React.createElement("div", null,
            React.createElement("label", null, "Basic Facepile:"),
            React.createElement(index_1.Facepile, { className: 'Facepile', personas: [{
                        imageUrl: './images/persona-female.png',
                        personaName: 'Annie Lindqvist',
                        imageInitials: 'AL',
                        data: '50%'
                    },
                    {
                        imageUrl: './images/persona-male.png',
                        personaName: 'Aaron Reid',
                        imageInitials: 'AR',
                        data: '$1,000'
                    }] }));
    };
    return FacepileVPage;
}(React.Component));
exports.default = FacepileVPage;
