// @microsoft/sp-loader/lib/test/mocks/MockSPComponentLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_api_1 = require("@microsoft/sp-core-library/test-api");
var MockSPComponentLoader = (function () {
    function MockSPComponentLoader() {
    }
    Object.defineProperty(MockSPComponentLoader.prototype, "manifests", {
        get: function () {
            return test_api_1.TEST_UNDEFINED;
        },
        enumerable: true,
        configurable: true
    });
    MockSPComponentLoader.prototype.startApplication = function (applicationId, preloadedData) {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype._initialize = function (preloadedData) {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.loadCss = function (url) {
        return;
    };
    MockSPComponentLoader.prototype.loadScript = function (url, options) {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.loadComponent = function (manifest) {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.loadComponentById = function (id, version) {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype._forceLocalManifests = function () {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.tryGetManifestById = function (id, version) {
        return test_api_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.registerManifests = function (manifests) {
        return;
    };
    return MockSPComponentLoader;
}());
exports.default = MockSPComponentLoader;
