// @microsoft/sp-module-interfaces/lib/manifestSchemas/examples/commandSet_1.manifest.js
"use strict";
var extensionManifest = {
    manifestVersion: 2,
    componentType: 'Extension',
    id: '00000000-0000-0000-0000-000000000000',
    alias: 'myExtension',
    version: '0.0.1',
    extensionType: 'ListViewCommandSet',
    items: {
        'item1': {
            title: {
                default: 'item 1 title'
            },
            type: 'command'
        },
        'item2': {
            title: {
                default: 'item 2 title',
                'en-us': 'Title'
            },
            type: 'command',
            ariaLabel: {
                default: 'title'
            },
            iconImageUrl: 'foo/bar.png'
        }
    },
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            'main.bundle': {
                type: 'path',
                path: 'bundle.script.js'
            }
        }
    }
};
module.exports = extensionManifest;
/* tslint:enable:export-name */
