// office-ui-fabric-react/lib-amd/components/Facepile/FacepilePage.visualtest.js
define(["require", "exports", "tslib", "./index", "react"], function (require, exports, tslib_1, index_1, React) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /* tslint:enable:no-unused-variable */
    var FacepileVPage = (function (_super) {
        tslib_1.__extends(FacepileVPage, _super);
        function FacepileVPage() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FacepileVPage.prototype.render = function () {
            return React.createElement("div", null,
                React.createElement("label", null, "Basic Facepile:"),
                React.createElement(index_1.Facepile, { className: 'Facepile', personas: [{
                            imageUrl: './images/persona-female.png',
                            personaName: 'Annie Lindqvist',
                            imageInitials: 'AL',
                            data: '50%'
                        },
                        {
                            imageUrl: './images/persona-male.png',
                            personaName: 'Aaron Reid',
                            imageInitials: 'AR',
                            data: '$1,000'
                        }] }));
        };
        return FacepileVPage;
    }(React.Component));
    exports.default = FacepileVPage;
});
