// @microsoft/sp-webpart-base/lib/test/mock/clientSideWebParts/MockWebPartContext.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var sp_core_library_1 = require("@microsoft/sp-core-library");
var WebPartContext_1 = require("../../../core/WebPartContext");
var MockWebPartManifestInstance_1 = require("./MockWebPartManifestInstance");
var MockWebPartContextParameters = (function () {
    function MockWebPartContextParameters() {
        this.domElement = document.createElement('div');
        this.manifest = new MockWebPartManifestInstance_1.default();
        this.instanceId = 'wpInstanceId';
        this.webPartTag = 'WebPart.daf0b71c-6de8-4ef7-b511-faae7c388708.0ae4973a-70e3-451e-9f1f-3f25e5455a60';
        this.loggingTag = 'WebPart.daf0b71c-6de8-4ef7-b511-faae7c388708.0ae4973a-70e3-451e-9f1f-3f25e5455a60';
        this.httpClient = {};
        this.spHttpClient = {};
        this.graphHttpClient = {};
        this.parentServiceScope = sp_core_library_1.ServiceScope.startNewRoot();
        this.statusRenderer = {};
        this.eventAggregator = {};
        this.host = {};
        this.propertyPane = {
            open: function () { },
            refresh: function () { }
        };
        this.isPropertyPaneRenderedByWebPart = function () { return true; };
        this.isPropertyPaneOpen = function () { return false; };
        this.setPropertyPaneState = function () { }; 
    }
    return MockWebPartContextParameters;
}());
exports.MockWebPartContextParameters = MockWebPartContextParameters;
var MockWebPartContext = (function (_super) {
    __extends(MockWebPartContext, _super);
    function MockWebPartContext(parameters) {
        return _super.call(this, parameters) || this;
    }
    MockWebPartContext.prototype.initialize = function (webPart) {
        var parentServiceScope = this.serviceScope.getParent();
        if (parentServiceScope) {
            parentServiceScope.finish();
        }
        webPart._internalInitialize(this);
        return webPart;
    };
    return MockWebPartContext;
}(WebPartContext_1.default));
exports.default = MockWebPartContext;
