// office-ui-fabric-react/lib-amd/components/CommandBar/CommandBarPage.visualtest.js
define(["require", "exports", "tslib", "./index", "react"], function (require, exports, tslib_1, index_1, React) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /* tslint:enable:no-unused-variable */
    var CommandBarVPage = (function (_super) {
        tslib_1.__extends(CommandBarVPage, _super);
        function CommandBarVPage() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        CommandBarVPage.prototype.render = function () {
            return React.createElement("div", { style: { width: '600px' } },
                React.createElement(index_1.CommandBar, { className: 'CommandBar', items: [
                        {
                            key: 'new',
                            name: 'New',
                            icon: 'Add',
                            onClick: function () { return; },
                            items: [
                                {
                                    key: 'emailMessage',
                                    name: 'Email message',
                                    icon: 'Mail'
                                },
                                {
                                    key: 'calendarEvent',
                                    name: 'Calendar event',
                                    icon: 'Calendar'
                                }
                            ]
                        },
                        {
                            key: 'upload',
                            name: 'Upload',
                            icon: 'Upload'
                        }
                    ] }));
        };
        return CommandBarVPage;
    }(React.Component));
    exports.default = CommandBarVPage;
});
