// office-ui-fabric-react/lib-amd/components/ComboBox/ComboBoxPage.visualtest.js
define(["require", "exports", "tslib", "./index", "react"], function (require, exports, tslib_1, index_1, React) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /* tslint:enable:no-unused-variable */
    var ComboBoxVPage = (function (_super) {
        tslib_1.__extends(ComboBoxVPage, _super);
        function ComboBoxVPage() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        ComboBoxVPage.prototype.render = function () {
            return React.createElement("div", null,
                React.createElement(index_1.ComboBox, { defaultSelectedKey: 'C', label: 'Basic uncontrolled example (allowFreeform: T, AutoComplete: T):', id: 'Basicdrop1', ariaLabel: 'Basic ComboBox example', allowFreeform: true, autoComplete: 'on', options: [
                        { key: 'A', text: 'Option a' },
                        { key: 'B', text: 'Option b' },
                        { key: 'C', text: 'Option c' },
                        { key: 'D', text: 'Option d' },
                        { key: 'E', text: 'Option e' },
                        { key: 'F', text: 'Option f' },
                        { key: 'G', text: 'Option g' },
                    ] }));
        };
        return ComboBoxVPage;
    }(React.Component));
    exports.default = ComboBoxVPage;
});
