// @uifabric/styling/lib-amd/styles/DefaultFontStyles.js
define(["require", "exports", "../glamorExports", "@uifabric/utilities/lib/language"], function (require, exports, glamorExports_1, language_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    // Default urls.
    var DefaultBaseUrl = 'https://static2.sharepointonline.com/files/fabric/assets';
    // Fallback fonts, if specified system or web fonts are unavailable.
    var FontFamilyFallbacks = "'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif";
    // Font face names to be registered.
    var FontNameArabic = 'Segoe UI Web (Arabic)';
    var FontNameCyrillic = 'Segoe UI Web (Cyrillic)';
    var FontNameEastEuropean = 'Segoe UI Web (East European)';
    var FontNameGreek = 'Segoe UI Web (Greek)';
    var FontNameHebrew = 'Segoe UI Web (Hebrew)';
    var FontNameThai = 'Leelawadee UI Web';
    var FontNameVietnamese = 'Segoe UI Web (Vietnamese)';
    var FontNameWestEuropean = 'Segoe UI Web (West European)';
    var FontNameSelawik = 'Selawik Web';
    // Font families with fallbacks, for the general regions.
    var FontFamilyArabic = "'" + FontNameArabic + "'";
    var FontFamilyChineseSimplified = "'Microsoft Yahei', Verdana, Simsun";
    var FontFamilyChineseTraditional = "'Microsoft Jhenghei', Pmingliu";
    var FontFamilyCyrillic = "'" + FontNameCyrillic + "'";
    var FontFamilyEastEuropean = "'" + FontNameEastEuropean + "'";
    var FontFamilyGreek = "'" + FontNameGreek + "'";
    var FontFamilyHebrew = "'" + FontNameHebrew + "'";
    var FontFamilyHindi = "'Nirmala UI'";
    var FontFamilyJapanese = "'Yu Gothic', 'Meiryo UI', Meiryo, 'MS Pgothic', Osaka";
    var FontFamilyKorean = "'Malgun Gothic', Gulim";
    var FontFamilySelawik = "'" + FontNameSelawik + "'";
    var FontFamilyThai = "'Leelawadee UI Web', 'Kmer UI'";
    var FontFamilyVietnamese = "'" + FontNameVietnamese + "'";
    var FontFamilyWestEuropean = "'" + FontNameWestEuropean + "'";
    // Mapping of language prefix to to font family.
    var LanguageToFontMap = {
        'ar': FontFamilyArabic,
        'bg': FontFamilyCyrillic,
        'cs': FontFamilyEastEuropean,
        'el': FontFamilyGreek,
        'et': FontFamilyEastEuropean,
        'he': FontFamilyHebrew,
        'hi': FontFamilyHindi,
        'hr': FontFamilyEastEuropean,
        'hu': FontFamilyEastEuropean,
        'ja': FontFamilyJapanese,
        'kk': FontFamilyEastEuropean,
        'ko': FontFamilyKorean,
        'lt': FontFamilyEastEuropean,
        'lv': FontFamilyEastEuropean,
        'pl': FontFamilyEastEuropean,
        'ru': FontFamilyCyrillic,
        'sk': FontFamilyEastEuropean,
        'sr-latn': FontFamilyEastEuropean,
        'th': FontFamilyThai,
        'tr': FontFamilyEastEuropean,
        'uk': FontFamilyCyrillic,
        'vi': FontFamilyVietnamese,
        'zh-hans': FontFamilyChineseSimplified,
        'zh-hant': FontFamilyChineseTraditional,
    };
    var FontFileVersion = 2.38;
    // Standard font sizes.
    var FontSizes;
    (function (FontSizes) {
        FontSizes.mini = '10px';
        FontSizes.xSmall = '11px';
        FontSizes.small = '12px';
        FontSizes.smallPlus = '13px';
        FontSizes.medium = '14px';
        FontSizes.mediumPlus = '15px';
        FontSizes.icon = '16px';
        FontSizes.large = '17px';
        FontSizes.xLarge = '21px';
        FontSizes.xxLarge = '28px';
        FontSizes.superLarge = '42px';
        FontSizes.mega = '72px';
    })(FontSizes = exports.FontSizes || (exports.FontSizes = {}));
    // Standard font weights.
    var FontWeights;
    (function (FontWeights) {
        FontWeights.light = 100;
        FontWeights.semilight = 300;
        FontWeights.regular = 400;
        FontWeights.semibold = 600;
        FontWeights.bold = 700;
    })(FontWeights = exports.FontWeights || (exports.FontWeights = {}));
    // Standard font styling.
    exports.DefaultFontStyles = {
        tiny: _createFont(FontSizes.mini, FontWeights.semibold),
        xSmall: _createFont(FontSizes.xSmall, FontWeights.regular),
        small: _createFont(FontSizes.small, FontWeights.regular),
        smallPlus: _createFont(FontSizes.smallPlus, FontWeights.regular),
        medium: _createFont(FontSizes.medium, FontWeights.regular),
        mediumPlus: _createFont(FontSizes.mediumPlus, FontWeights.regular),
        large: _createFont(FontSizes.large, FontWeights.semilight),
        xLarge: _createFont(FontSizes.xLarge, FontWeights.light),
        xxLarge: _createFont(FontSizes.xxLarge, FontWeights.light),
        superLarge: _createFont(FontSizes.superLarge, FontWeights.light),
        mega: _createFont(FontSizes.mega, FontWeights.light),
        icon: {
            fontFamily: '"FabricMDL2Icons"',
            fontWeight: FontWeights.regular,
            fontStyle: 'normal'
        }
    };
    function _getFontFamily() {
        var language = language_1.getLanguage();
        var fontFamily = FontFamilyWestEuropean;
        for (var lang in LanguageToFontMap) {
            if (LanguageToFontMap.hasOwnProperty(lang) && language && lang.indexOf(language) === 0) {
                // tslint:disable-next-line:no-any
                fontFamily = LanguageToFontMap[lang];
                break;
            }
        }
        return fontFamily + ", " + FontFamilyFallbacks;
    }
    function _createFont(size, weight) {
        return {
            fontFamily: _getFontFamily(),
            MozOsxFontSmoothing: 'grayscale',
            WebkitFontSmoothing: 'antialiased',
            fontSize: size,
            fontWeight: weight
        };
    }
    function _registerFontFace(fontFamily, url, fontWeight) {
        fontFamily = "'" + fontFamily + "'";
        glamorExports_1.fontFace({
            fontFamily: fontFamily,
            src: "url('" + url + ".woff2') format('woff2')," +
                ("url('" + url + ".woff') format('woff')"),
            fontWeight: fontWeight,
            fontStyle: 'normal'
        });
    }
    function _registerFontFaceSet(baseUrl, fontFamily, cdnFolder, cdnFontName) {
        if (cdnFontName === void 0) { cdnFontName = 'segoeui'; }
        var urlBase = baseUrl + "/" + cdnFolder + "/" + cdnFontName;
        _registerFontFace(fontFamily, urlBase + '-light', FontWeights.light);
        _registerFontFace(fontFamily, urlBase + '-semilight', FontWeights.semilight);
        _registerFontFace(fontFamily, urlBase + '-regular', FontWeights.regular);
        _registerFontFace(fontFamily, urlBase + '-semibold', FontWeights.semibold);
    }
    function _registerDefaultFontFaces() {
        var baseUrl = _getFontBaseUrl();
        if (baseUrl) {
            var fontUrl = baseUrl + "/fonts";
            var iconUrl = baseUrl + "/icons";
            // Produce @font-face definitions for all supported web fonts.
            _registerFontFaceSet(fontUrl, FontNameThai, 'leelawadeeui-thai', 'leelawadeeui');
            _registerFontFaceSet(fontUrl, FontNameArabic, 'segoeui-arabic');
            _registerFontFaceSet(fontUrl, FontNameCyrillic, 'segoeui-cyrillic');
            _registerFontFaceSet(fontUrl, FontNameEastEuropean, 'segoeui-easteuropean');
            _registerFontFaceSet(fontUrl, FontNameGreek, 'segoeui-greek');
            _registerFontFaceSet(fontUrl, FontNameHebrew, 'segoeui-hebrew');
            _registerFontFaceSet(fontUrl, FontNameVietnamese, 'segoeui-vietnamese');
            _registerFontFaceSet(fontUrl, FontNameWestEuropean, 'segoeui-westeuropean');
            _registerFontFaceSet(fontUrl, FontFamilySelawik, 'selawik', 'selawik');
            // Leelawadee UI (Thai) does not have a 'light' weight, so we override
            // the font-face generated above to use the 'semilight' weight instead.
            _registerFontFace('Leelawadee UI Web', fontUrl + "/leelawadeeui-thai/leelawadeeui-semilight", FontWeights.light);
            // Leelawadee UI (Thai) does not have a 'semibold' weight, so we override
            // the font-face generated above to use the 'bold' weight instead.
            _registerFontFace('Leelawadee UI Web', fontUrl + "/leelawadeeui-thai/leelawadeeui-bold", FontWeights.semibold);
            // Register icon urls.
            _registerFontFace('FabricMDL2Icons', iconUrl + "/fabricmdl2icons-" + FontFileVersion, FontWeights.regular);
        }
    }
    /**
     * Reads the fontBaseUrl from window.FabricConfig.fontBaseUrl or falls back to a default.
     */
    function _getFontBaseUrl() {
        var win = typeof window !== 'undefined' ? window : undefined;
        // tslint:disable-next-line:no-string-literal no-any
        var fabricConfig = win ? win['FabricConfig'] : undefined;
        return (fabricConfig && fabricConfig.fontBaseUrl !== undefined) ? fabricConfig.fontBaseUrl : DefaultBaseUrl;
    }
    /**
     * Register the font faces.
     */
    _registerDefaultFontFaces();
});
