// office-ui-fabric-react/lib/components/DocumentCard/DocumentCardPage.visualtest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var index_1 = require("./index");
/* tslint:disable:no-unused-variable */
var React = require("react");
/* tslint:enable:no-unused-variable */
var DocumentCardVPage = (function (_super) {
    tslib_1.__extends(DocumentCardVPage, _super);
    function DocumentCardVPage() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DocumentCardVPage.prototype.render = function () {
        var previewProps = {
            previewImages: [
                {
                    name: 'DocumentCard',
                    previewImageSrc: 'dist/document-preview.png',
                    iconSrc: 'dist/icon-ppt.png',
                    width: 318,
                    height: 196
                }
            ],
        };
        return React.createElement("div", null,
            React.createElement(index_1.DocumentCard, { className: 'DocumentCard' },
                React.createElement(index_1.DocumentCardPreview, tslib_1.__assign({}, previewProps)),
                React.createElement(index_1.DocumentCardTitle, { title: 'Document Card', shouldTruncate: true }),
                React.createElement(index_1.DocumentCardActivity, { activity: 'Created a few minutes ago', people: [
                        { name: 'Annie Lindqvist', profileImageSrc: 'images/persona-female.png' }
                    ] })));
    };
    return DocumentCardVPage;
}(React.Component));
exports.default = DocumentCardVPage;
