// @microsoft/sp-build-core-tasks/lib/configureWebpack/ConfigureExternalBundlingWebpackTask.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var lodash = require("lodash");
var path = require("path");
var ExternalsProcessor_1 = require("./ExternalsProcessor");
var OdspGulpTask_1 = require("./../OdspGulpTask");
var CopyAssetsTask_1 = require("./../copyAssets/CopyAssetsTask");
/**
 * Configures the @microsoft/gulp-core-build-webpack task to bundle localized externals into another, pre-built package.
 *
 * Example:
 *  IN:
 *    setConfig({
 *      webpack: <reference to webpack task>,
 *      localizedResources: <optional - list of localized externals>
 *      configureWebpackTask: <reference to the ConfigureWebpackTask task>,
 *      debugLocale: <optional - locale to use for a debug build>,
 *      additionalLoaders: <optional - additional loaders for localized resources>
 *    })
 *
 *  OUT:
 *    If any resources are specified and found, set webpack config to:
 *      {
 *        entry: <entries with locales>,
 *        module: {
 *          rules: ([
 *          ] <and additional loaders, if specified>,
 *          noParse: [ /\.map$/ ],
 *        },
 *        resolve: {
 *          alias: <localized resources>,
 *          modulesDirectories: [
 *            '',
 *            'node_modules',
 *            'lib'
 *          ]
 *        },
 *        context: <dist directory>,
 *        devtool: 'source-map',
 *        externals: <externals and non-specified locales>,
 *        entry: webpackEntries,
 *        output: {
 *          libraryTarget: <library target from configureWebpackTask, if specified> or 'amd',
 *          path: this.outputDirectory
 *        },
 *        plugins: []
 *      }
 */
var ConfigureExternalBundlingWebpackTask = (function (_super) {
    __extends(ConfigureExternalBundlingWebpackTask, _super);
    function ConfigureExternalBundlingWebpackTask() {
        return _super.call(this, 'configure-webpack-external-bundling', {
            webpack: undefined,
            localizedResources: undefined,
            bundleAllLocalizedResources: false
        }) || this;
    }
    ConfigureExternalBundlingWebpackTask.prototype.loadSchema = function () {
        return require('./configure-webpack-external-bundling.schema.json');
    };
    ConfigureExternalBundlingWebpackTask.prototype.setConfig = function (config) {
        if (config.webpack && config.webpack.name !== 'webpack') {
            this.logWarning('Setting webpack property to non-"@microsoft/gulp-core-build-webpack"-type will ' +
                'prevent configuration');
        }
        _super.prototype.mergeConfig.call(this, config);
    };
    ConfigureExternalBundlingWebpackTask.prototype.executeTask = function (gulp, completeCallback) {
        if (!this.taskConfig.webpack) {
            this.logWarning('"webpack" is not defined in the task configuration. Nothing to configure.');
            completeCallback();
            return;
        }
        if (!this.taskConfig.configureWebpackTask) {
            this.logError('"configureWebpackTask" is not defined in the task configuration.');
            this._setWebpackConfig(undefined);
            completeCallback();
            return;
        }
        if (!this._shouldBeEnabled) {
            this.logVerbose('No localized resources specified, nothing to do.');
            this._setWebpackConfig(undefined);
            completeCallback();
            return;
        }
        if (!this.properties.webpackStats) {
            this.logError('Missing webpack stats object. Unable to configure.');
            this._setWebpackConfig(undefined);
            completeCallback();
            return;
        }
        var resourcesToBundle = this._pruneLocalizedResources();
        if (resourcesToBundle.length === 0) {
            this.logVerbose('None of the specified localized resources were present in the webpack stats object. Nothing ' +
                'to do.');
            this._setWebpackConfig(undefined);
            completeCallback();
            return;
        }
        // When we run webpack again, it'll overwrite this.properties.webpackStats, and we want to remember the
        //  webpack stats object from the first webpack run.
        this.properties.initialWebpackStats = this.properties.webpackStats;
        var resolveResult;
        try {
            resolveResult = this.resolveResources(resourcesToBundle, this.taskConfig.debugLocale, false);
        }
        catch (error) {
            completeCallback(error);
            return;
        }
        var locales = resolveResult.locales;
        var localizedResourcesToBundle = resolveResult.localizedResourcesToBundle;
        var externalsKeys = ExternalsProcessor_1.getExternalsKeys(this);
        var localizedResources = this.properties.localizedResources || {};
        var localizedResourcesKeys = Object.keys(localizedResources).filter(function (key) {
            return localizedResourcesToBundle.indexOf(key) === -1;
        });
        if (!this.properties.bundledExternals) {
            this.properties.bundledExternals = [];
        }
        (_a = this.properties.bundledExternals).push.apply(_a, localizedResourcesToBundle);
        var webpackEntries = this._generateWebpackEntries(locales);
        var config = {
            module: {
                rules: [
                    {
                        use: require.resolve('source-map-loader'),
                        enforce: 'pre',
                        test: /\.js$/
                    }
                ].concat(this.taskConfig.additionalLoaders || []),
                noParse: [/\.map$/]
            },
            resolve: {
                alias: {},
                modules: [
                    'node_modules',
                    'lib'
                ]
            },
            context: this.outputDirectory,
            devtool: this.buildConfig.production ? undefined : 'source-map',
            externals: externalsKeys.concat(localizedResourcesKeys),
            entry: webpackEntries,
            output: {
                library: this.properties.webpackLibraryName,
                libraryTarget: 'amd',
                path: this.outputDirectory,
                devtoolModuleFilenameTemplate: 'webpack:///../[resource-path]',
                devtoolFallbackModuleFilenameTemplate: 'webpack:///../[resource-path]?[hash]'
            }
        };
        // VSO#258841: Minification here doesn't make a big deal (<1kb gzipped), but adds a lot of build time.
        //  This can updated eventually to drop a fake base bundle, run webpack, and then replace the fake bundle
        //  with the real payload.
        // if (this.buildConfig.production) {
        //   config.plugins.push(new webpack.optimize.UglifyJsPlugin({
        //     compress: {
        //       dead_code: false,
        //       warnings: false
        //     },
        //     mangle: true
        //   }));
        // }
        this.taskConfig.configureWebpackTask.tinkerWithLocConfig(config);
        var configs = [];
        for (var _i = 0, locales_1 = locales; _i < locales_1.length; _i++) {
            var locale = locales_1[_i];
            var localeWebpackConfig = lodash.cloneDeep(config);
            for (var _b = 0, localizedResourcesToBundle_1 = localizedResourcesToBundle; _b < localizedResourcesToBundle_1.length; _b++) {
                var resourceName = localizedResourcesToBundle_1[_b];
                var resourcePath = this.properties.discoveredLocalizedResources[resourceName][locale];
                localeWebpackConfig.resolve.alias[resourceName] = resourcePath;
            }
            localeWebpackConfig.output.filename = "[name]_" + locale + ".js";
            localeWebpackConfig.output.chunkFilename = "[id].[name]_" + locale + "_[chunkhash].js";
            configs.push(localeWebpackConfig);
        }
        this._setWebpackConfig(configs);
        completeCallback();
        var _a;
    };
    /**
     * Modify the base webpack config to make sure we get what we need after that webpack instance runs.
     *
     * @internal
     */
    ConfigureExternalBundlingWebpackTask.prototype.tinkerWithBaseWebpackConfig = function (config) {
        if (this._shouldBeEnabled) {
            config.output.libraryTarget = 'amd';
        }
    };
    ConfigureExternalBundlingWebpackTask.prototype._generateWebpackEntries = function (locales) {
        var _this = this;
        var entries = this.properties.bundleEntries || [];
        var webpackEntries = {};
        entries.forEach(function (entry) {
            var entryName = path.basename(entry.nonLocalizedEntrypointPath, '.js');
            webpackEntries[entryName] = path.join(_this.outputDirectory, entryName + ".js");
            entry.nonLocalizedEntrypointPath = undefined;
            entry.localizedEntrypointPaths = {};
            locales.forEach(function (locale) {
                entry.localizedEntrypointPaths[locale] = path.join(_this.outputDirectory, entryName + "_" + locale + ".js");
            });
        });
        return webpackEntries;
    };
    ConfigureExternalBundlingWebpackTask.prototype._pruneLocalizedResources = function () {
        var modules = this.properties.webpackStats.toJson({ modules: true }).modules;
        var filteredResources = [];
        for (var _i = 0, modules_1 = modules; _i < modules_1.length; _i++) {
            var module_1 = modules_1[_i];
            var matches = module_1.name.match(CopyAssetsTask_1.externalsRegex);
            if (matches) {
                var externalName = matches[1];
                if ((this.taskConfig.bundleAllLocalizedResources &&
                    this.properties.discoveredLocalizedResources[externalName]) ||
                    (this.taskConfig.localizedResources && this.taskConfig.localizedResources.indexOf(externalName) !== -1)) {
                    filteredResources.push(externalName);
                }
            }
        }
        return filteredResources;
    };
    Object.defineProperty(ConfigureExternalBundlingWebpackTask.prototype, "_shouldBeEnabled", {
        get: function () {
            return this.taskConfig.configureWebpackTask &&
                (this.taskConfig.bundleAllLocalizedResources &&
                    this.taskConfig.configureWebpackTask.properties.discoveredLocalizedResources &&
                    Object.keys(this.taskConfig.configureWebpackTask.properties.discoveredLocalizedResources).length !== 0) ||
                (this.taskConfig.localizedResources &&
                    this.taskConfig.localizedResources.length !== 0);
        },
        enumerable: true,
        configurable: true
    });
    ConfigureExternalBundlingWebpackTask.prototype._setWebpackConfig = function (config) {
        var webpackTask = this.taskConfig.webpack;
        if (webpackTask) {
            if (!config) {
                // We're just going to disable the second webpack task
                webpackTask.isEnabled = function () { return false; };
            }
            else {
                webpackTask.setConfig({
                    config: config,
                    configPath: null // tslint:disable-line:no-null-keyword
                });
            }
        }
    };
    return ConfigureExternalBundlingWebpackTask;
}(OdspGulpTask_1.default));
exports.ConfigureExternalBundlingWebpackTask = ConfigureExternalBundlingWebpackTask;
