// office-ui-fabric-react/lib-amd/components/Button/DefaultButton/DefaultButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities", "../BaseButton.styles", "../SplitButton/SplitButton.styles"], function (require, exports, Styling_1, Utilities_1, BaseButton_styles_1, SplitButton_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DEFAULT_BUTTON_HEIGHT = '32px';
    var DEFAULT_BUTTON_MINWIDTH = '80px';
    var DEFAULT_PADDING = '0 16px';
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles, focusInset, focusColor) {
        var baseButtonStyles = BaseButton_styles_1.getStyles(theme, focusInset, focusColor);
        var splitButtonStyles = SplitButton_styles_1.getStyles(theme);
        var defaultButtonStyles = {
            root: {
                minWidth: DEFAULT_BUTTON_MINWIDTH,
                height: DEFAULT_BUTTON_HEIGHT,
                backgroundColor: theme.palette.neutralLighter,
                color: theme.palette.neutralPrimary
            },
            rootHovered: {
                backgroundColor: theme.palette.neutralLight,
                color: theme.palette.black
            },
            rootPressed: {
                backgroundColor: theme.palette.neutralTertiaryAlt,
                color: theme.palette.neutralDark
            },
            rootExpanded: {
                backgroundColor: theme.palette.neutralTertiaryAlt,
                color: theme.palette.neutralDark
            },
            rootChecked: {
                backgroundColor: theme.palette.neutralTertiaryAlt,
                color: theme.palette.neutralDark
            },
            label: {
                fontWeight: Styling_1.FontWeights.semibold
            }
        };
        return Styling_1.mergeStyleSets(baseButtonStyles, defaultButtonStyles, splitButtonStyles, customStyles);
    });
});
