// @microsoft/sp-webpart-base/lib/components/animations/_Animations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _Animations = (function () {
    function _Animations() {
        this._pendingElements = [];
    }
    Object.defineProperty(_Animations, "instance", {
        get: function () {
            if (!this._instance) {
                this._instance = new _Animations();
            }
            return this._instance;
        },
        enumerable: true,
        configurable: true
    });
    _Animations.prototype.init = function () {
        var _this = this;
        this._ensureScrollRevealLoaded().then(function () {
            _this._pendingElements.forEach(function (elem) {
                _this._scrollReveal.reveal(elem);
            });
        });
    };
    _Animations.prototype.reveal = function (elem) {
        if (!!this._scrollReveal) {
            this._scrollReveal.reveal(elem);
        }
        else {
            this._pendingElements.push(elem);
        }
    };
    _Animations.prototype._ensureScrollRevealLoaded = function () {
        var _this = this;
        return new Promise(function (resolve) {
            if (!_this._scrollReveal) {
                require.ensure([], function (require) {
                    var sr = require('scrollReveal');
                    _this._scrollReveal =
                        sr({
                            scale: 1,
                            origin: 'bottom',
                            duration: 750,
                            distance: '20px',
                            easing: 'cubic-bezier(0.33, 0.0, 0.0, 1)',
                            container: 'div[class*="scrollRegion_"]'
                        });
                    resolve(_this._scrollReveal);
                });
            }
            else {
                resolve(_this._scrollReveal);
            }
        });
    };
    return _Animations;
}());
exports.default = _Animations;
