// @microsoft/sp-component-base/lib/test/BaseComponentContext.test.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var BaseComponentContext_1 = require("../BaseComponentContext");
var TestComponent = (function (_super) {
    __extends(TestComponent, _super);
    function TestComponent(parameters) {
        return _super.call(this, parameters) || this;
    }
    return TestComponent;
}(BaseComponentContext_1.default));
var rootScope = sp_core_library_1.ServiceScope.startNewRoot();
rootScope.finish();
var testParameters = {
    parentServiceScope: rootScope,
    manifest: {},
    loggingTag: 'TestComponent',
    instanceId: 'TestInstanceId'
};
describe('BaseComponentContext', function () {
    it('BaseComponentContext should be testable', function (done) {
        var testComponent = new TestComponent(testParameters);
        chai_1.assert.isOk(testComponent);
        done();
    });
});
