// @microsoft/api-extractor/lib/DebugRun.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
// NOTE: THIS SOURCE FILE IS FOR DEBUGGING PURPOSES ONLY.
//       IT IS INVOKED BY THE "Run.cmd" AND "Debug.cmd" BATCH FILES.
var ts = require("typescript");
var path = require("path");
var os = require("os");
var Extractor_1 = require("./Extractor");
var ApiFileGenerator_1 = require("./generators/ApiFileGenerator");
var ApiJsonGenerator_1 = require("./generators/ApiJsonGenerator");
var ROOT_DIR = './test/inputs/example4';
var ENTRY_POINT = 'src/index.ts';
var compilerOptions = {
    target: ts.ScriptTarget.ES5,
    module: ts.ModuleKind.CommonJS,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    jsx: ts.JsxEmit.React,
    rootDir: ROOT_DIR,
    types: ['es6-collections', 'webpack-env']
};
var extractor = new Extractor_1.default({
    compilerOptions: compilerOptions,
    errorHandler: function (message, fileName, lineNumber) {
        console.log("ErrorHandler: " + message + os.EOL
            + ("  " + fileName + "#" + lineNumber));
    }
});
extractor.loadExternalPackages('./testInputs/external-api-json');
process.chdir(ROOT_DIR);
extractor.analyze({
    entryPointFile: path.join(ROOT_DIR, ENTRY_POINT),
    otherFiles: [
        path.join(ROOT_DIR, './typings/tsd.d.ts')
    ]
});
// Normally warnings are kept by the ApiItem data structure,
// and written to the '*.api.ts' file.
var warnings = [];
var apiFileGenerator = new ApiFileGenerator_1.default();
apiFileGenerator.writeApiFile(path.join(__dirname, './DebugRun-Output.api.ts'), extractor);
var apiJsonGenerator = new ApiJsonGenerator_1.default();
apiJsonGenerator.writeJsonFile(path.join(__dirname, './DebugRun-Output.json'), extractor);
console.log('DebugRun completed.');
