// @uifabric/styling/lib/styles/getFocusStyle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("../utilities/index");
var glamorExports_1 = require("../glamorExports");
/**
 * Generates a focus style which can be used to define an :after focus border.
 *
 * @param theme - The theme object to use.
 * @param inset - The number of pixels to inset the border.
 * @param color - The color for the border.
 * @param position - The positioning applied to the container. Must
 * be 'relative' or 'absolute' so that the focus border can live around it.
 * @returns The style object.
 */
function getFocusStyle(theme, inset, color, position) {
    if (inset === void 0) { inset = '0'; }
    if (color === void 0) { color = theme.palette.neutralSecondary; }
    if (position === void 0) { position = 'relative'; }
    return index_1.mergeStyles({
        outline: 'transparent',
        '::-moz-focus-inner': { border: 0 },
        position: position
    }, glamorExports_1.parent('.ms-Fabric.is-focusVisible', {
        ':focus:after': {
            content: '""',
            position: 'absolute',
            left: inset,
            top: inset,
            bottom: inset,
            right: inset,
            border: '1px solid ' + color
        }
    }));
}
exports.getFocusStyle = getFocusStyle;
