// @uifabric/styling/lib/styles/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AnimationStyles_1 = require("./AnimationStyles");
exports.AnimationStyles = AnimationStyles_1.AnimationStyles;
var DefaultPalette_1 = require("./DefaultPalette");
exports.DefaultPalette = DefaultPalette_1.DefaultPalette;
var DefaultFontStyles_1 = require("./DefaultFontStyles");
exports.DefaultFontStyles = DefaultFontStyles_1.DefaultFontStyles;
exports.FontSizes = DefaultFontStyles_1.FontSizes;
exports.FontWeights = DefaultFontStyles_1.FontWeights;
var IconCodes_1 = require("./IconCodes");
exports.IconCodes = IconCodes_1.IconCodes;
var getFocusStyle_1 = require("./getFocusStyle");
exports.getFocusStyle = getFocusStyle_1.getFocusStyle;
var UtilityStyles_1 = require("./UtilityStyles");
exports.hideText = UtilityStyles_1.hideText;
var theme_1 = require("./theme");
exports.ThemeSettingName = theme_1.ThemeSettingName;
exports.getTheme = theme_1.getTheme;
exports.loadTheme = theme_1.loadTheme;
exports.createTheme = theme_1.createTheme;
