// office-ui-fabric-react/lib-amd/components/Button/CommandButton/CommandButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities", "../BaseButton.styles", "../SplitButton/SplitButton.styles"], function (require, exports, Styling_1, Utilities_1, BaseButton_styles_1, SplitButton_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DEFAULT_BUTTON_HEIGHT = '40px';
    var DEFAULT_PADDING = '0 4px';
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
        var baseButtonStyles = BaseButton_styles_1.getStyles(theme);
        var splitButtonStyles = SplitButton_styles_1.getStyles(theme);
        var commandButtonStyles = {
            root: {
                borderWidth: '0',
                padding: DEFAULT_PADDING,
                height: DEFAULT_BUTTON_HEIGHT,
                color: theme.palette.neutralPrimary,
                backgroundColor: 'transparent'
            },
            rootHovered: {
                color: theme.palette.themeDarker
            },
            rootPressed: {
                color: theme.palette.themePrimary
            },
            rootDisabled: {
                color: theme.palette.neutralTertiary,
                backgroundColor: 'transparent'
            },
            rootChecked: {
                backgroundColor: theme.palette.neutralTertiaryAlt,
            },
            rootExpanded: {
                color: theme.palette.themePrimary
            },
            rootCheckedHovered: {
                backgroundColor: theme.palette.neutralLight
            },
            flexContainer: {
                justifyContent: 'flex-start'
            },
            icon: {
                color: theme.palette.themePrimary
            },
            iconDisabled: {
                color: 'inherit'
            },
            menuIcon: {
                color: theme.palette.neutralSecondary
            }
        };
        return Styling_1.mergeStyleSets(baseButtonStyles, commandButtonStyles, splitButtonStyles, customStyles);
    });
});
