// office-ui-fabric-react/lib/components/Button/CompoundButton/CompoundButton.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Styling_1 = require("../../../Styling");
var Utilities_1 = require("../../../Utilities");
var DefaultButton_styles_1 = require("../DefaultButton/DefaultButton.styles");
var SplitButton_styles_1 = require("../SplitButton/SplitButton.styles");
var DEFAULT_BUTTON_HEIGHT = '32px';
var DEFAULT_BUTTON_MINWIDTH = '80px';
var DEFAULT_PADDING = '0 16px';
exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
    var defaultButtonStyles = DefaultButton_styles_1.getStyles(theme, customStyles);
    var splitButtonStyles = SplitButton_styles_1.getStyles(theme);
    var compoundButtonStyles = {
        root: {
            maxWidth: '280px',
            minHeight: '72px',
            height: 'auto',
            padding: '20px'
        },
        flexContainer: {
            flexDirection: 'column',
            alignItems: 'flex-start',
            minWidth: '100%',
            margin: ''
        },
        label: {
            margin: '0 0 5px',
            lineHeight: '100%'
        },
        description: [
            theme.fonts.small,
            {
                color: theme.palette.neutralSecondary,
                lineHeight: '100%'
            }
        ],
        descriptionHovered: {
            color: theme.palette.neutralDark
        },
        descriptionPressed: {
            color: 'inherit'
        },
        descriptionChecked: {
            color: 'inherit'
        },
        descriptionDisabled: {
            color: 'inherit'
        }
    };
    return Styling_1.mergeStyleSets(defaultButtonStyles, compoundButtonStyles, splitButtonStyles, customStyles);
});
