// office-ui-fabric-react/lib-amd/components/CommandBar/examples/CommandBar.Example.scss.js
define(["require", "exports", "@microsoft/load-themed-styles"], function (require, exports, load_themed_styles_1) {
    "use strict";
    var styles = {
        customButtonContainer: 'customButtonContainer_0007a2a4',
        icon: 'icon_0007a2a4',
        themeDarkAltColor: 'themeDarkAltColor_0007a2a4',
        leftText: 'leftText_0007a2a4',
        splitter: 'splitter_0007a2a4',
        button: 'button_0007a2a4',
        darkerBG: 'darkerBG_0007a2a4',
    };
    load_themed_styles_1.loadStyles([{ "rawString": ".customButtonContainer_0007a2a4{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-align:center;-ms-flex-align:center;align-items:center}.icon_0007a2a4{display:inline-block;font-family:FabricMDL2Icons;font-style:normal;font-weight:400;speak:none;font-size:14px;position:relative}.themeDarkAltColor_0007a2a4{color:" }, { "theme": "themeDarkAlt", "defaultValue": "#106ebe" }, { "rawString": "}.leftText_0007a2a4{margin-left:6px;font-size:14px;font-weight:400;color:" }, { "theme": "neutralPrimary", "defaultValue": "#333333" }, { "rawString": "}.splitter_0007a2a4{font-size:14px;font-weight:400;color:" }, { "theme": "neutralTertiary", "defaultValue": "#a6a6a6" }, { "rawString": ";background:0 0;position:relative;top:-2px;height:40px}.button_0007a2a4.ms-Button{height:40px;min-width:28px;padding:0 5px}.button_0007a2a4:hover{background:initial}.darkerBG_0007a2a4{background:#d0d0d0}" }]);
    return styles;
});
/* tslint:enable */
