// office-ui-fabric-react/lib/components/Checkbox/examples/Checkbox.Basic.Example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var Checkbox_1 = require("office-ui-fabric-react/lib/Checkbox");
var Persona_1 = require("office-ui-fabric-react/lib/Persona");
var CheckboxBasicExample = (function (_super) {
    tslib_1.__extends(CheckboxBasicExample, _super);
    function CheckboxBasicExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            isChecked: false
        };
        _this._onCheckboxChange = _this._onCheckboxChange.bind(_this);
        return _this;
    }
    CheckboxBasicExample.prototype.render = function () {
        var _this = this;
        var isChecked = this.state.isChecked;
        var styles = {
            root: {
                marginTop: '10'
            }
        };
        return (React.createElement("div", null,
            React.createElement(Checkbox_1.Checkbox, { label: 'Uncontrolled checkbox', onChange: this._onCheckboxChange, inputProps: {
                    onFocus: function () { console.log('Uncontrolled checkbox is focused'); },
                    onBlur: function () { console.log('Uncontrolled checkbox is blured'); }
                }, styles: styles }),
            React.createElement(Checkbox_1.Checkbox, { label: 'Uncontrolled checkbox with defaultChecked true', defaultChecked: true, onChange: this._onCheckboxChange, styles: styles }),
            React.createElement(Checkbox_1.Checkbox, { label: 'Disabled uncontrolled checkbox', disabled: true, styles: styles }),
            React.createElement(Checkbox_1.Checkbox, { label: 'Disabled uncontrolled checkbox with defaultChecked true', disabled: true, defaultChecked: true, onChange: this._onCheckboxChange, styles: styles }),
            React.createElement(Checkbox_1.Checkbox, { label: 'Controlled checkbox', checked: isChecked, onChange: function (ev, checked) {
                    _this.setState({ isChecked: checked });
                }, styles: styles }),
            React.createElement(Checkbox_1.Checkbox, { label: 'Checkbox rendered with boxSide "end"', boxSide: 'end', styles: styles }),
            React.createElement(Checkbox_1.Checkbox, { label: 'Persona Checkbox', styles: styles, onRenderLabel: this._renderPersonaLabel })));
    };
    CheckboxBasicExample.prototype._onCheckboxChange = function (ev, isChecked) {
        console.log("The option has been changed to " + isChecked + ".");
    };
    CheckboxBasicExample.prototype._renderPersonaLabel = function (props) {
        var DEFAULT_IMAGE_URL = '/_layouts/15/userphoto.aspx?size=S&accountname=';
        return React.createElement(Persona_1.Persona, { primaryText: props.label, imageUrl: DEFAULT_IMAGE_URL, size: Persona_1.PersonaSize.extraSmall });
    };
    return CheckboxBasicExample;
}(React.Component));
exports.CheckboxBasicExample = CheckboxBasicExample;
